### PR TITLE
Enemy ai

### PR DESCRIPTION
**Pull request by TahriT**
1. I removed enemyfixed  and enemyrandom from the levels. 
2. Only enemyPatrol. 
3. If you don't set a position the enemy is fixed 
4. Shoots and agros only on LOS 
5. Has levels between 1 and 5 which tweaks some parameters 
6. Lose agro after certain time without LOS 
7. Enemy animation are attached to agent speed .

### CODE REVIEW

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 1]
```diff
@@ -62,13 +150,35 @@ public int GetDamage()
...
+
+            // Set the agent to go to the currently selected destination.
+            agent.destination = transform.position;
+
+            // Choose the next point in the array as the destination,
+            // cycling to the start if necessary.
+            // destPoint = (destPoint + 1) % points.Length;
+        }
+
+        public void TargetPlayer()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: Make abstract if you want this to be empty

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 2]
```diff
@@ -62,13 +150,35 @@ public int GetDamage()
...
+        {
+            //override
+        }
+
+        public void Chase()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: Make abstract if you want this to be empty

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 3]
```diff
@@ -62,13 +150,35 @@ public int GetDamage()
             return this.damage;
         }
 
-        protected void Shoot()
+        public void GotoNextPoint()
+        {
+            // Returns if no points have been set up
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: why are these comments here?

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 4]
```diff
@@ -5,43 +5,131 @@
...
+                    cooldown = 3f;
+                    fov = 110f;
+                    agent.speed += 1.5f;
+                    agroRange = 3f;
+                    break;
+
+            }
+        }
+
+        private void FixedUpdate()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: why is this here if its empty

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 5]
```diff
@@ -100,14 +211,19 @@ private void OnCollisionEnter(Collision collision)
             }
         }
 
+        // Handels player attacks on enemy
         private void OnTriggerEnter(Collider other)
         {
+
+            //Depricated Moving to On Trigger Stay with Raycast
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: why is this here? if something is deprecated remove it

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 6]
```diff
@@ -129,13 +245,54 @@ private void OnTriggerEnter(Collider other)
...
+                    else
+                    {
+                        los = false;
+                        if (((interest) -= Time.deltaTime) <= 0 && !los)
+                        {
+                             agro = false;
+                             GotoNextPoint();
+                            //lose interest in the player
+                        }
+
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: remove this extra empty space

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 7]
```diff
@@ -62,13 +150,35 @@ public int GetDamage()
...
+        {
+            //override
+        }
+
+        public void Chase()
+        {
+            //override
+        }
+
+        //Shoots
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: use xml comments for method summaries ex. /// \<summary> \</summary>

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 8]
```diff
@@ -88,6 +197,8 @@ protected void Shoot()
             }
         }
 
+
+        // fixed update on raycasts
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: why is this comment here?

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 9]
```diff
@@ -100,14 +211,19 @@ private void OnCollisionEnter(Collision collision)
             }
         }
 
+        // Handels player attacks on enemy
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: don't put method summaries on Unity methods, put this comment near the code that does this

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 10]
```diff
@@ -129,13 +245,54 @@ private void OnTriggerEnter(Collider other)
             }
         }
 
+        //Active while in range of player
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: don't put method summaries on Unity methods, put this comment near the code that does this

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 11]
```diff
@@ -129,13 +245,54 @@ private void OnTriggerEnter(Collider other)
...
+
+                if (Physics.Raycast(this.gunPos.transform.position, other.transform.position - this.gunPos.transform.position, out hit) && hit.transform.tag == "Player")
+                {
+                    if (hit.collider.tag == "Player")
+                    {
+                        InAgroRange();
+                        this.agro = true;
+                        this.InAgroRange();
+                        los = true;
+                        //agro animation
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs:233]: what does this comment mean?

File: Assets/Scripts/Character/Enemy/EnemyAI.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+ 
+        public EnemyAI(Enemy enemy)
+        {
+            this.enemy = enemy;
+        }      
+
+        /*
+         Find random location within navmesh
+         */
+        public Vector3 RandomNavmeshLocation(float radius)
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyAI.cs]: use xml comments for method summaries ex. /// \<summary> \</summary>

File: Assets/Scripts/Character/Enemy/EnemyAI.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            if (NavMesh.SamplePosition(randomDirection, out hit, radius, 1))
+            {
+                finalPosition = hit.position;
+            }
+            return finalPosition;
+        }
+        /*
+        Find random location within navmesh from origin point
+        */
+        public static Vector3 RandomNavSphere(Vector3 origin, float dist, int layermask)
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyAI.cs]: use xml comments for method summaries ex. /// \<summary> \</summary>

File: Assets/Scripts/Character/Enemy/EnemyAI.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+            NavMesh.SamplePosition(randDirection, out navHit, dist, layermask);
+
+            return navHit.position;
+        }
+
+       
+
+        // Use this for initialization
+        void Start()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyAI.cs]: remove this if its empty

File: Assets/Scripts/Character/Enemy/EnemyAI.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            // define the level of the enemy
+        
+        }
+
+        // Update is called once per frame
+        void Update()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyAI.cs]: remove this if its empty

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs [hunk 16]
```diff
@@ -27,77 +25,70 @@ void GotoNextPoint()
...
-            Vector3 finalPosition = Vector3.zero;
-            if (NavMesh.SamplePosition(randomDirection, out hit, radius, 1))
-            {
-                finalPosition = hit.position;
-            }
-            return finalPosition;
         }
-
-        void attack()
+        // Chase player
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: use xml comments for method summaries ex. /// \<summary> \</summary>

File: Assets/Scripts/Character/Enemy/EnemyAI.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
+﻿namespace GGJ2018.Character.Enemy.Smarts
+{
+    using System.Collections;
+    using System.Collections.Generic;
+    using UnityEngine;
+    using UnityEngine.AI;
+    using GGJ2018.Managers;
+
+    public class EnemyAI : MonoBehaviour
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyAI.cs:9]: what is this class supposed to do?  if its an AI parent then make it extend Enemy.cs

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs [hunk 18]
```diff
@@ -1,24 +1,22 @@
...
-        private float fov = 90f;
-        public float cooldown = 5f;
-
-        void GotoNextPoint()
+        public EnemyAI smarts; //for future use with position variation
+        public Transform[] points; // locations to patr
+        public float reset;  // reset for cooldown
+        public float intrest;
+        
+        new public void GotoNextPoint()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: don't use 'new'.  If you want to override a method make the method in the parent class virtual then add the keyword override to this.

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs [hunk 19]
```diff
@@ -27,77 +25,70 @@ void GotoNextPoint()
...
-            if (NavMesh.SamplePosition(randomDirection, out hit, radius, 1))
-            {
-                finalPosition = hit.position;
-            }
-            return finalPosition;
         }
-
-        void attack()
+        // Chase player
+       new  public void Chase()
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: don't use 'new'.  If you want to override a method make the method in the parent class virtual then add the keyword override to this.

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs [hunk 20]
```diff
@@ -27,77 +25,70 @@ void GotoNextPoint()
...
         {
-            enemyAnimator.SetBool("isMoving", true);
-            Vector3 dist = PlayerManager.Instance.player.position - agent.transform.position;
-            if (dist.magnitude > agroRange)
+            //Chase the player to the agroRange distance which is the minimum distance needed to shoot
+            Vector3 dist = PlayerManager.Instance.player.position -  agent.transform.position;
+            if (dist.magnitude >  agroRange)
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: remove extra space

File: Assets/Scripts/Character/Player/Player.cs [hunk 21]
```diff
@@ -26,7 +26,7 @@ public class Player : MonoBehaviour
         public HealthUI healthUI;
         public Transform gunPos;
         public BoxCollider CaneHitbox;
-
+
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Player/Player.cs:29]: why is this file change here?  don't commit a change unless there is actual content

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 22]
```diff
@@ -5,43 +5,131 @@
...
+        public float fov = 90f;
+        public float interest = 0;
+        private float hate = 15f;
+        private bool smart = true;
+ 
 
         private void Start()
         {
-          
+
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: remove extra line

File: Assets/Scripts/Character/Enemy/Enemy.cs [hunk 23]
```diff
@@ -5,43 +5,131 @@
...
+                    agent.speed += 1.25f;
+                    agroRange = 4f;
+                    break;
+                case 5:
+                    cooldown = 3f;
+                    fov = 110f;
+                    agent.speed += 1.5f;
+                    agroRange = 3f;
+                    break;
+
```
Comment (JonathanHunter) [inline Assets/Scripts/Character/Enemy/Enemy.cs]: remove extra line

**Review by JonathanHunter — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 4 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Though it was needed for ray-casting to work properly something you mentioned before. Removing

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 11 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs:233]: Place holder for enemy animation from Larry

File: Assets/Scripts/Character/Enemy/EnemyAI.cs (hunk 12 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyAI.cs]: Fixed https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyAI.cs (hunk 13 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyAI.cs]: Fixed https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyAI.cs (hunk 14 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyAI.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyAI.cs (hunk 15 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyAI.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs (hunk 18 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: Fixed but please check again
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs (hunk 16 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/EnemyPatrol.cs (hunk 19 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/EnemyPatrol.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Player/Player.cs (hunk 21 above)
Comment (TahriT) [reply Assets/Scripts/Character/Player/Player.cs:29]: My bad. Was looking for example code of yours. >.<

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 22 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 23 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 4 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 3 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 1 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed I think made it virtual https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 2 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed I think I made it virtual  
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 7 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 8 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 9 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 5 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

File: Assets/Scripts/Character/Enemy/Enemy.cs (hunk 10 above)
Comment (TahriT) [reply Assets/Scripts/Character/Enemy/Enemy.cs]: Fixed
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

**Comment by TahriT** [comment]
I made the changes but I've got merge conflicts. Not sure what happened. But please review again.
https://github.com/JonathanHunter/GGJ2018/pull/3/commits/7bbc53c24cf9c4d3eb1a7a23ae98e730651850f1

**Review by JonathanHunter — APPROVED** [review]

**Pull request merged**